src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,473 @@
+
+            var currentNode = token.Parent;
+            // if cursor is right before a closing delimiter, make sure you start with node outside of delimiters
+            if ((token.Kind() == SyntaxKind.OpenBraceToken || token.Kind() == SyntaxKind.OpenBracketToken || token.Kind() == SyntaxKind.OpenParenToken) && token.Span.End == caret.Value.Position)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (GetEnclosingArgumentList(currentNode, token, caret, caretPosition, syntaxFacts) == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var lastDelimiterPosition = -1;
+            var finalDelimiterNeedsSemicolon = false;
+
+            // work your way out, verifying all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!ReachedSemicolonSyntax(currentNode, syntaxFacts, ref finalDelimiterNeedsSemicolon))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode, ref lastDelimiterPosition))
+                {
+                    nextCommandHandler();
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    nextCommandHandler();
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode.Ancestors().Any(n => n.Kind() == SyntaxKind.ForStatement))
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (finalDelimiterNeedsSemicolon)
+            {
+                if (!StatementClosingDelimiterExists(currentNode, ref lastDelimiterPosition))
+                {
+                    nextCommandHandler();
+                    return;
+                }
+            }
+
+            // if you haven't found any enclosures, put semicolon at end of statement
+            if (lastDelimiterPosition < 0)
+            {
+                lastDelimiterPosition = currentNode.Span.End;
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterPosition, currentNode.Kind())));
+            nextCommandHandler();
+        }
+
+        private SyntaxNode GetEnclosingArgumentList(SyntaxNode currentNode, SyntaxToken token, SnapshotPoint? caret, int caretPosition, ISyntaxFactsService syntaxFacts)
+        {
+            if (token.Kind() == SyntaxKind.OpenParenToken && caretPosition == currentNode.SpanStart)
```

Comment:
💡 This condition is actually handled by the caller. If you remove the duplicate handling from this method, you can also remove the `token`, `caret`, and `caretPosition` parameters.